website/mails/mails.py

```diff
@@ -117,6 +124,79 @@ def send_mail(to_addr, mail, mimetype='plain', from_addr=None, mailer=None,
 
         return ret
 
+def queue_mail(to_addr, mail, send_at, user, **context):
+    """
+    Queue an email to be sent using send_mail after a specified amount
+    of time and if the callback returns True. The callback is attached to
+    the template under mail.
+
+    :param to_addr: the address email is to be sent to
+    :param mail:  the type of mail. Struct following template:
+                        { 'callback': function(),
+                            'template': mako template name,
+                            'subject': mail subject }
+    :param send_at: datetime object of when to send mail
+    :param user: user object attached to mail
+    :param context: IMPORTANT kwargs to be attached to template.
+                    Sending mail will fail if wrong all kwargs are
+                    not parameters.
+    :return: the QueuedMail object created
+    """
+    new_mail = QueuedMail(
+        user=user,
+        to_addr=to_addr,
+        send_at=send_at,
+        email_type=mail['template'],
+        data=context
+    )
+    new_mail.save()
+    return new_mail
+
+class QueuedMail(StoredObject):
+    _id = fields.StringField(primary=True, default=lambda: str(bson.ObjectId()))
+    user = fields.ForeignField('User')
+    to_addr = fields.StringField()
+    send_at = fields.DateTimeField()
+    email_type = fields.StringField()
+    data = fields.DictionaryField()
+    sent_at = fields.DateTimeField(index=True)
+
+    def send_mail(self):
+        """
+        Grabs the data from this email, checks for user subscription to help mails,
+
+        constructs the mail object and checks callback. Then attempts to send the email
+        through send_mail()
+        :return: boolean based on whether email was sent.
+        """
+        mail_struct = queue_mail_types[self.email_type]
+        callback = mail_struct['callback'](self)
+        mail = Mail(
+            mail_struct['template'],
+            subject=mail_struct['subject']
+        )
+        self.data['osf_url'] = settings.DOMAIN
+        if callback and self.user.osf_mailing_lists.get(settings.OSF_HELP_LIST):
+            send_mail(self.to_addr, mail, mimetype='html', **(self.data or {}))
+            self.sent_at = datetime.utcnow()
+            self.save()
+            return True
+        else:
+            self.__class__.remove_one(self)
+            return False
+
+    def find_same_email_sent_to_same_user(self):
+        """
+        Queries up for all emails of the same type as self, sent to the same user as self.
+        Does not look for queue-up emails.
+        :return: a list of those emails
+        """
+        return list(self.__class__.find(
```

Comment:
You shouldn't need to coerce to a list here. I would expect any `.find_*` methods to return a queryset. It will be the responsibility of the caller to cast to a list if necessary.